File: LeetCode/LeetCode 81. 搜索旋转排序数组 II (二分).cpp
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int l = 0, r = nums.size() - 1;
        while(l <= r) {
            int mid = ((l ^ r) >> 1) + (l & r);
            if(nums[mid] == target) return true;
            if(nums[l] == nums[mid]) l++;
            else if(nums[l] < nums[mid]) {
                if(nums[l] <= target && nums[mid] > target) r = mid - 1;
                else l = mid + 1;
            }
            else {
                if(nums[r] >= target && nums[mid] < target) l = mid + 1;
                else r = mid - 1;
            }
        }
        return false;
    }
};

File: LeetCode/LeetCode 81. 搜索旋转排序数组 II (二分).py
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        l, r = 0, len(nums) - 1

        while l <= r:
            mid = (l + r) // 2
            if nums[mid] == target:
                return True
            if nums[l] == nums[mid]:
                l += 1
            elif nums[l] < nums[mid]:
                if nums[l] <= target and nums[mid] > target:
                    r = mid - 1
                else:
                    l = mid + 1
            else:
                if nums[r] >= target and nums[mid] < target:
                    l = mid + 1
                else:
                    r = mid - 1

        return False 
